crates/types/src/consensus.rs

```diff
@@ -549,27 +553,39 @@ impl<TYPES: NodeType> Consensus<TYPES> {
     ///
     /// Returns true if the action is for a newer view than the last action of that type
     pub fn update_action(&mut self, action: HotShotAction, view: TYPES::View) -> bool {
-        let old_view = match action {
-            HotShotAction::Vote => &mut self.last_actions.voted,
-            HotShotAction::Propose => &mut self.last_actions.proposed,
-            HotShotAction::DaPropose => &mut self.last_actions.da_proposed,
+        match action {
             HotShotAction::DaVote => {
+                // Use vote tracker to prevent double voting
+                let voter_key = Arc::new(self.public_key().clone());
+                
+                if !self.vote_tracker.record_vote(view, voter_key) {
+                    tracing::warn!("Prevented double voting attempt for view {}", view);
+                    return false;
+                }
+
                 if view > self.last_actions.da_vote {
                     self.last_actions.da_vote = view;
                 }
-                // TODO Add logic to prevent double voting.  For now the simple check if
-                // the last voted view is less than the view we are trying to vote doesn't work
-                // because the leader of view n + 1 may propose to the DA (and we would vote)
-                // before the leader of view n.
-                return true;
+                
+                // Clean up old vote records periodically
+                self.vote_tracker.cleanup_old_views(view);
```

Comment:
Doesn't this open us up tot he opposite problem would have? Vote in view `n`, then view `n+1` then we can vote for view `n` again?  The old code of course just allows any duplicate vote so there is an improvement with this code.

Reply:
@bfish713 I have implemented a robust system for tracking view sequences. Added a queue mechanism that handles out-of-order votes. This prevents double voting while allowing the network to be resilient. Sorry, I missed the importance of the TODO comment. Comprehensive test coverage of both epoch transitions and network partitioning scenarios is included in this solution.